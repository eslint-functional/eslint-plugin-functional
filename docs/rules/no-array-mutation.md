### no-array-mutation

[![Type Info Required][type-info-badge]][type-info-url]

This rule prohibits mutating an array via assignment to or deletion of their elements/properties. This rule enforces array immutability without the use of `ReadonlyArray<T>` (as apposed to [readonly-array](#readonly-array)).

```typescript
const x = [0, 1, 2];

x[0] = 4; // <- Mutating an array is not allowed.
x.length = 1; // <- Mutating an array is not allowed.
x.push(3); // <- Mutating an array is not allowed.
```

#### Has Fixer

No

#### Options

- [ignore-prefix](#using-the-ignore-prefix-option)
- [ignore-suffix](#using-the-ignore-suffix-option)
- [ignore-pattern](#using-the-ignore-pattern-option)
- [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)
- ~~ignore-mutation-following-accessor~~ - _deprecated in favor of [ignore-new-array](#using-the-ignore-new-array-option-with-no-array-mutation)_

#### Example config

```javascript
"no-array-mutation": true
```

```javascript
"no-array-mutation": [true, {"ignore-prefix": "mutable"}]
```

```javascript
"no-array-mutation": [true, "ignore-new-array"]
```
