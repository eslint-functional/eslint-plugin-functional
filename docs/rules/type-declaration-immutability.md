# Enforce a level of immutability for type declaration (type-declaration-immutability)

Require type alias declarations and interfaces that imply some level of
immutability to comply to it.

## Rule Details

This rule enforces rules on type immutability based on the type's name.

For details on what the different levels of immutability mean, see [the
immutability
section](https://github.com/RebeccaStevens/is-immutable-type#immutability) of
[is-immutable-type](https://www.npmjs.com/package/is-immutable-type).

### ❌ Incorrect

<!-- eslint-disable functional/type-declaration-immutability -->

```ts
/* eslint functional/type-declaration-immutability: "error" */

type ReadonlyElement = {
  id: number;
  data: string[];
};

type ReadonlyDeepElement = Readonly<{
  id: number;
  data: string[];
}>;

type MutableElement = Readonly<{
  id: number;
  data: ReadonlyArray<string>;
}>;
```

### ✅ Correct

<!-- eslint-disable functional/type-declaration-immutability -->

```ts
/* eslint functional/type-declaration-immutability: "error" */

type ReadonlyElement = Readonly<{
  id: number;
  data: string[];
}>;

type ReadonlyDeepElement = Readonly<{
  id: number;
  data: ReadonlyArray<string>;
}>;

type MutableElement = {
  readonly id: number;
  data: ReadonlyArray<string>;
};
```

## Settings

This rule can leverage shared settings to configure immutability settings.

See the [immutability](./settings/immutability.md) docs.

## Options

This rule accepts an options object of the following type:

```ts
type Options = {
  rules: Array<{
    identifiers: string | string[];
    immutability:  "Mutable" | "ReadonlyShallow" | "ReadonlyDeep" | "Immutable";
    comparator?: "Less" | "AtMost" | "Exactly" | "AtLeast" | "More";
  }>;
  ignoreInterfaces: boolean;
  ignorePattern: string[] | string;
}
```

### Default Options

```ts
const defaults = {
  rules: [
    {
      identifiers: "^(?!I?Mutable).+",
      immutability: "Immutable",
      comparator: "AtLeast",
    },
  ],
  ignoreInterfaces: false,
}
```

### Preset Overrides

#### `recommended` and `lite`

```ts
const recommendedOptions = {
  rules: [
    {
      identifiers: "I?Immutable.+",
      immutability: "Immutable",
      comparator: "AtLeast",
    },
    {
      identifiers: "I?ReadonlyDeep.+",
      immutability: "ReadonlyDeep",
      comparator: "AtLeast",
    },
    {
      identifiers: "I?Readonly.+",
      immutability: "ReadonlyShallow",
      comparator: "AtLeast",
    },
    {
      identifiers: "I?Mutable.+",
      immutability: "Mutable",
      comparator: "AtMost",
    },
  ],
}
```

### `rules`

An array of rules to enforce immutability by.

These rules should be sorted by precedence as each type declaration will only
enforce the first matching rule to it.

#### `identifiers`

A regex pattern or an array of regex patterns that are used to match against the
name of the type declarations.

#### `immutability`

The level of immutability to compare against. This value will be compared to the
calculated immutability using the `comparator`.

#### `comparator`

The comparator to use to compare the calculated immutability to the desired
immutability. This can be thought of as `<`, `<=`, `==`, `>=` or `>`.

### `ignoreInterfaces`

A boolean to specify whether interfaces should be exempt from these rules.
`false` by default.

### `ignorePattern`

See the [ignorePattern](./options/ignore-pattern.md) docs.
