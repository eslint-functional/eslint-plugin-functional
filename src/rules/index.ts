import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClass from "./no-class";
import * as noConditionalStatement from "./no-conditional-statement";
import * as noExpressionStatement from "./no-expression-statement";
import * as noLet from "./no-let";
import * as noLoop from "./no-loop-statement";
import * as noMethodSignature from "./no-method-signature";
import * as noMixedType from "./no-mixed-type";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpression from "./no-this-expression";
import * as noThrowStatement from "./no-throw-statement";
import * as noTryStatement from "./no-try-statement";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferType from "./prefer-type-literal";

/**
 * All of the custom rules.
 */
export const rules = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClass.name]: noClass.rule,
  [noConditionalStatement.name]: noConditionalStatement.rule,
  [noExpressionStatement.name]: noExpressionStatement.rule,
  [noLet.name]: noLet.rule,
  [noLoop.name]: noLoop.rule,
  [noMethodSignature.name]: noMethodSignature.rule,
  [noMixedType.name]: noMixedType.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpression.name]: noThisExpression.rule,
  [noThrowStatement.name]: noThrowStatement.rule,
  [noTryStatement.name]: noTryStatement.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferType.name]: preferType.rule,
};
