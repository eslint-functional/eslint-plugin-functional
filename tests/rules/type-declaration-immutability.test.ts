import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { Immutability } from "is-immutable-type";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/type-declaration-immutability";

import { typescriptConfig } from "../utils/configs";

describe(name, () => {
  describe("typescript", () => {
    const options = [
      {
        rules: [
          {
            identifiers: ["^I?Immutable.+"],
            immutability: Immutability.Immutable,
            comparator: "AtLeast",
          },
          {
            identifiers: ["^I?ReadonlyDeep.+"],
            immutability: Immutability.ReadonlyDeep,
            comparator: "AtLeast",
          },
          {
            identifiers: ["^I?Readonly.+"],
            immutability: Immutability.ReadonlyShallow,
            comparator: "AtLeast",
          },
          {
            identifiers: ["^I?Mutable.+"],
            immutability: Immutability.Mutable,
            comparator: "AtMost",
          },
        ],
      },
    ];

    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: typescriptConfig,
    });

    it("reports invalid shallow records", async () => {
      const invalidResult1 = await invalid({
        code: dedent`
          type ReadonlyFoo = {
            foo: number;
          };
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult1.result).toMatchSnapshot();

      const invalidResult2 = await invalid({
        code: dedent`
          type ReadonlyFoo = {
            readonly foo: number;
            bar: {
              baz: string;
            };
          }
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult2.result).toMatchSnapshot();
    });

    it("reports invalid immutable/deep records", async () => {
      const invalidResult1 = await invalid({
        code: dedent`
          type ReadonlyDeepFoo = {
                      readonly foo: number;
                      readonly bar: {
                        baz: string;
                      };
                    };
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult1.result).toMatchSnapshot();

      const invalidResult2 = await invalid({
        code: dedent`
          type ImmutableFoo = {
                      readonly foo: number;
                      readonly bar: {
                        baz: string;
                      };
                    };
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult2.result).toMatchSnapshot();
    });

    it("reports invalid shallow arrays", async () => {
      const invalidResult = await invalid({
        code: `type ReadonlyMyArray = Array<string>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid deep arrays", async () => {
      const invalidResult = await invalid({
        code: `type ReadonlyDeepMyArray = ReadonlyArray<{ foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid immutable arrays", async () => {
      const invalidResult = await invalid({
        code: `type ImmutableMyArray = ReadonlyArray<{ readonly foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid shallow sets", async () => {
      const invalidResult = await invalid({
        code: `type ReadonlyMySet = Set<string>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid deep sets", async () => {
      const invalidResult = await invalid({
        code: `type ReadonlyDeepMySet = ReadonlySet<{ foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid immutable sets", async () => {
      const invalidResult = await invalid({
        code: `type ImmutableMySet = ReadonlySet<{ readonly foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid shallow maps", async () => {
      const invalidResult = await invalid({
        code: `type ReadonlyMyMap = Map<string, string>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid deep maps", async () => {
      const invalidResult = await invalid({
        code: `type ReadonlyDeepMyMap = ReadonlyMap<string, { foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports invalid immutable maps", async () => {
      const invalidResult = await invalid({
        code: `type ImmutableMyMap = ReadonlyMap<string, { readonly foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports non-mutable primitives", async () => {
      const invalidResult = await invalid({
        code: "type MutableString = string;",
        options,
        errors: ["AtMost"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports non-mutable records", async () => {
      const invalidResult1 = await invalid({
        code: dedent`
          type MutableFoo = {
            readonly foo: number;
          };
        `,
        options,
        errors: ["AtMost"],
      });
      expect(invalidResult1.result).toMatchSnapshot();

      const invalidResult2 = await invalid({
        code: dedent`
          type MutableFoo = Readonly<{
            foo: number;
          }>;
        `,
        options,
        errors: ["AtMost"],
      });
      expect(invalidResult2.result).toMatchSnapshot();
    });

    it("doesn't report mutable sets", async () => {
      await valid({
        code: `type MutableSet = Set<string>;`,
        options,
      });
    });

    it("doesn't report mutable map", async () => {
      await valid({
        code: `type MutableMap = Map<string, string>;`,
        options,
      });
    });

    it("doesn't report valid types", async () => {
      await valid({
        code: "type ReadonlyString = string;",
        options,
      });
      await valid({
        code: "type ReadonlyFoo = { readonly foo: number };",
        options,
      });
      await valid({
        code: "type ReadonlyFoo = Readonly<{ foo: number }>;",
        options,
      });
      await valid({
        code: "type ReadonlyFoo = { readonly foo: number; readonly bar: { baz: string; }; };",
        options,
      });
      await valid({
        code: "type ReadonlySet = ReadonlySet<string>;",
        options,
      });
      await valid({
        code: "type ReadonlyMap = ReadonlyMap<string, string>;",
        options,
      });
      await valid({
        code: "type ReadonlyDeepString = string;",
        options,
      });
      await valid({
        code: "type ReadonlyDeepFoo = { readonly foo: number; readonly bar: { readonly baz: string; }; };",
        options,
      });
      await valid({
        code: "type ReadonlyDeepSet = ReadonlySet<string>;",
        options,
      });
      await valid({
        code: "type ReadonlyDeepMap = ReadonlyMap<string, string>;",
        options,
      });
      await valid({
        code: "type ReadonlyDeepSet = ReadonlySet<{ readonly foo: string; }>;",
        options,
      });
      await valid({
        code: "type ReadonlyDeepMap = ReadonlyMap<string, { readonly foo: string; }>;",
        options,
      });
      await valid({
        code: "type ImmutableString = string;",
        options,
      });
      await valid({
        code: "type ImmutableFoo = { readonly foo: number; readonly bar: { readonly baz: string; }; };",
        options,
      });
      await valid({
        code: "type ImmutableSet = Readonly<ReadonlySet<{ readonly foo: string; }>>;",
        options,
      });
      await valid({
        code: "type ImmutableMap = Readonly<ReadonlyMap<string, { readonly foo: string; }>>;",
        options,
      });
      await valid({
        code: "type MutableFoo = { foo: number };",
        options,
      });
      await valid({
        code: "type MutableFoo = { readonly foo: number; bar: { readonly baz: string; }; };",
        options,
      });
      await valid({
        code: "type MutableSet = Set<{ readonly foo: string; }>;",
        options,
      });
      await valid({
        code: "type MutableMap = Map<string, { readonly foo: string; }>;",
        options,
      });
    });

    describe("ignoreInterfaces", () => {
      it("doesn't report interfaces when enabled", async () => {
        await valid({
          code: dedent`
            interface ReadonlyFoo {
              foo: number;
            }
          `,
          options: [{ ignoreInterfaces: true }],
        });
      });
    });

    describe("ignoreIdentifierPattern", () => {
      it("doesn't report interfaces when enabled", async () => {
        await valid({
          code: dedent`
            type ReadonlyFoo = {
              foo: number;
            };
          `,
          options: [
            {
              ignoreIdentifierPattern: "Foo",
            },
          ],
        });
      });
    });

    it("respects override settings", async () => {
      await valid({
        code: dedent`
          type ReadonlyDeepFoo = ReadonlyDeep<{ foo: { bar: string; }; }>;
          type ReadonlyDeep<T> = T;
        `,
        options,
        settings: {
          immutability: {
            overrides: [
              {
                type: "ReadonlyDeep",
                to: Immutability.ReadonlyDeep,
              },
            ],
          },
        },
      });

      const invalidResult = await invalid({
        code: `type MutableSet = Set<string>;`,
        options,
        settings: {
          immutability: {
            overrides: {
              keepDefault: false,
              values: [
                {
                  type: { from: "lib", name: "Set" },
                  to: Immutability.Immutable,
                },
              ],
            },
          },
        },
        errors: ["AtMost"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });
  });
});
