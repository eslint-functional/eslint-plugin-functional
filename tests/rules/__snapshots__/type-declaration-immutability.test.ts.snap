// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`type-declaration-immutability > typescript > reports invalid deep arrays 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 25,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyDeepMyArray = ReadonlyArray<{ foo: string; }>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 25,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyDeepMyArray = ReadonlyArray<{ foo: string; }>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid deep maps 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 23,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyDeepMyMap = ReadonlyMap<string, { foo: string; }>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 23,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyDeepMyMap = ReadonlyMap<string, { foo: string; }>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid deep sets 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 23,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyDeepMySet = ReadonlySet<{ foo: string; }>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 23,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyDeepMySet = ReadonlySet<{ foo: string; }>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable arrays 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 22,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ImmutableMyArray = ReadonlyArray<{ readonly foo: string; }>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 22,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ImmutableMyArray = ReadonlyArray<{ readonly foo: string; }>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable maps 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 20,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ImmutableMyMap = ReadonlyMap<string, { readonly foo: string; }>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 20,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ImmutableMyMap = ReadonlyMap<string, { readonly foo: string; }>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable sets 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 20,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ImmutableMySet = ReadonlySet<{ readonly foo: string; }>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 20,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ImmutableMySet = ReadonlySet<{ readonly foo: string; }>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable/deep records 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 21,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyDeepFoo = {
            readonly foo: number;
            readonly bar: {
              baz: string;
            };
          };",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 21,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyDeepFoo = {
            readonly foo: number;
            readonly bar: {
              baz: string;
            };
          };",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable/deep records 2`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 18,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyShallow").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ImmutableFoo = {
            readonly foo: number;
            readonly bar: {
              baz: string;
            };
          };",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 18,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyShallow").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ImmutableFoo = {
            readonly foo: number;
            readonly bar: {
              baz: string;
            };
          };",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow arrays 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 21,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyMyArray = Array<string>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 21,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyMyArray = Array<string>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow maps 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 19,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyMyMap = Map<string, string>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 19,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyMyMap = Map<string, string>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow records 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 17,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyFoo = {
  foo: number;
};",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 17,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyFoo = {
  foo: number;
};",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow records 2`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 17,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyFoo = {
  readonly foo: number;
  bar: {
    baz: string;
  };
}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 17,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyFoo = {
  readonly foo: number;
  bar: {
    baz: string;
  };
}",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow sets 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 19,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "AtLeast",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type ReadonlyMySet = Set<string>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 19,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "AtLeast",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type ReadonlyMySet = Set<string>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports non-mutable primitives 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 19,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
      "messageId": "AtMost",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type MutableString = string;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 19,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
          "messageId": "AtMost",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type MutableString = string;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports non-mutable records 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 16,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
      "messageId": "AtMost",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type MutableFoo = {
  readonly foo: number;
};",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 16,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
          "messageId": "AtMost",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type MutableFoo = {
  readonly foo: number;
};",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > reports non-mutable records 2`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 16,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
      "messageId": "AtMost",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type MutableFoo = Readonly<{
  foo: number;
}>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 16,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
          "messageId": "AtMost",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type MutableFoo = Readonly<{
  foo: number;
}>;",
    },
  ],
}
`;

exports[`type-declaration-immutability > typescript > respects override settings 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 6,
      "endColumn": 16,
      "endLine": 1,
      "line": 1,
      "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
      "messageId": "AtMost",
      "nodeType": "Identifier",
      "ruleId": "type-declaration-immutability",
      "severity": 2,
    },
  ],
  "output": "type MutableSet = Set<string>;",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 6,
          "endColumn": 16,
          "endLine": 1,
          "line": 1,
          "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
          "messageId": "AtMost",
          "nodeType": "Identifier",
          "ruleId": "rule-to-test/type-declaration-immutability",
          "severity": 2,
        },
      ],
      "output": "type MutableSet = Set<string>;",
    },
  ],
}
`;
